Userland/Libraries/LibCore/FileWatcherLinux.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2023, Tim Flynn <trflynn89@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "FileWatcher.h"
+#include <AK/Debug.h>
+#include <AK/DeprecatedString.h>
+#include <AK/LexicalPath.h>
+#include <Kernel/API/InodeWatcherFlags.h>
+#include <LibCore/Notifier.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/inotify.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#if !defined(AK_OS_LINUX)
+static_assert(false, "This file must only be used for Linux");
+#endif
+
+namespace Core {
+
+static constexpr unsigned inode_watcher_flags_to_inotify_flags(InodeWatcherFlags flags)
+{
+    unsigned result = 0;
+
+    if ((flags & InodeWatcherFlags::Nonblock) != InodeWatcherFlags::None)
+        result |= IN_NONBLOCK;
+    if ((flags & InodeWatcherFlags::CloseOnExec) != InodeWatcherFlags::None)
+        result |= IN_CLOEXEC;
+
+    return result;
+}
+
+static Optional<FileWatcherEvent> get_event_from_fd(int fd, HashMap<unsigned, DeprecatedString> const& wd_to_path)
+{
+    static constexpr auto max_event_size = sizeof(inotify_event) + NAME_MAX + 1;
+
+    // Note from INOTIFY(7) man page:
+    //
+    //     Some systems cannot read integer variables if they are not properly aligned. On other
+    //     systems, incorrect alignment may decrease performance. Hence, the buffer used for reading
+    //     from the inotify file descriptor should have the same alignment as inotify_event.
+    alignas(alignof(inotify_event)) Array<u8, max_event_size> buffer;
+    ssize_t rc = ::read(fd, buffer.data(), buffer.size());
+
+    if (rc == 0) {
+        return {};
+    } else if (rc < 0) {
+        dbgln_if(FILE_WATCHER_DEBUG, "get_event_from_fd: Reading from wd {} failed: {}", fd, strerror(errno));
+        return {};
+    }
+
+    auto const* event = reinterpret_cast<inotify_event const*>(buffer.data());
+    FileWatcherEvent result;
+
+    auto it = wd_to_path.find(event->wd);
+    if (it == wd_to_path.end()) {
+        dbgln_if(FILE_WATCHER_DEBUG, "get_event_from_fd: Got an event for a non-existent wd {}?!", event->wd);
+        return {};
+    }
+
+    auto const& path = it->value;
+
+    if ((event->mask & IN_CREATE) != 0) {
+        result.type = FileWatcherEvent::Type::ChildCreated;
+    } else if ((event->mask & IN_DELETE) != 0) {
+        result.type = FileWatcherEvent::Type::ChildDeleted;
+    } else if ((event->mask & IN_DELETE_SELF) != 0) {
+        result.type = FileWatcherEvent::Type::Deleted;
+    } else if ((event->mask & IN_MODIFY) != 0) {
+        result.type = FileWatcherEvent::Type::ContentModified;
+    } else if ((event->mask & IN_ATTRIB) != 0) {
```

Comment:
Is it possible for the event mask to have multiple bits set?

Reply:
Seems so. I patched that and made a couple commits to handle this mask having multiple events.